### PR TITLE
Fix crash in SymInt unary minus

### PR DESCRIPTION
**Pull request by malfet**
Before this change `-SymInt(std::numeric_limits<int64_t>::min()) == 0` would reliably crash with null pointer dereference, as `data_` of the SymInt returned by `operator-` would be `0x8000000000000000`, because of the carry/overflow flags set by `negq`. 

Before the change x86_64 assembly generated for 
https://github.com/pytorch/pytorch/blob/4f02cc06706648799a33218711e62eae859d5927/c10/core/SymInt.cpp#L137
looked as follows:
```
   0x7ffff7f2f490 <+115>: movq   %rax, %rdx
    0x7ffff7f2f493 <+118>: negq   %rdx
    0x7ffff7f2f496 <+121>: movq   %rdx, (%rbp)
    0x7ffff7f2f49a <+125>: movabsq $0x4000000000000000, %rdx ; imm = 0x4000000000000000 
    0x7ffff7f2f4a4 <+135>: cmpq   %rdx, %rax
    0x7ffff7f2f4a7 <+138>: jle    0x7ffff7f2f520            ; <+259> at SymInt.cpp:141:1
```
`negq %rfx` correspond to unary minus and  `cmpq   %rdx, 0x4000000000000000` are inverted `check_range` 
https://github.com/pytorch/pytorch/blob/b6d0d0819ae6099b2d659d03892e1e19c17e912b/c10/core/SymInt.h#L247-L249
Flags raised by `negq` will affect the results of `cmpq`, and as result value would not be allocated on heap, but rather preserved as `nullptr`. 

Not sure if it's worth benchmarking, but perhaps using `__builtin_sub_overflow` would be faster as it does not require an extra comparison, just guarantees that overflow flags is cleared after the op.

### CODE REVIEW

**Review by Skylion007 — APPROVED** [review]

**Review by colesbury — APPROVED** [review]

File: c10/core/SymInt.cpp [hunk 1]
```diff
@@ -134,7 +134,13 @@ bool SymInt::expect_size(const char* file, int64_t line) const {
 
 SymInt operator-(const SymInt& s) {
   if (auto ma = s.maybe_as_int()) {
-    return SymInt(-*ma);
+    const auto val = *ma;
+    // Note: Result of `-std::numeric_limits<decltype(val)>::min()` is undefined
+    // But on many platforms it equals to self + setting Carry/Overflow flags
+    // Which in opimized code affects results of `check_range` condition
```
Comment (albanD) [inline c10/core/SymInt.cpp:140]:
```suggestion
    // Which in optimized code affects results of `check_range` condition
```

**Review by albanD — COMMENTED** [review]
Can you check the perf hit?
You can use Timer with cpp code to measure the difference on the neg code:
```python
import torch
from timeit import default_timer

global_setup = """
#include <c10/core/SymInt.h>
"""
setup = """
c10::SymInt a = c10::SymInt(123);
"""
code = """
-a;
"""

from torch.utils.benchmark import Timer

t = Timer(stmt=code, setup=setup, global_setup=global_setup, language="c++", timer=default_timer)

print(t.blocked_autorange())
```
You can vary the value to hit either case of course.

**Comment by malfet** [comment]
@albanD thank you for the benchmark, run it on Intel(R) Xeon(R) Platinum 8275CL CPU @ 3.00GHz
Before the change it takes `4.19 ns` after `4.44 ns`, i.e. 3% slow down. Will try to `__builtin_sub_overflow` a bit later

**Comment by ezyang** [comment]
Very nice, thanks!

**Pull request closed without merging**